src/elfpy/pricing_models/yieldspace.py

```diff
@@ -110,134 +110,141 @@ def calc_lp_out_given_tokens_in(
-        assert (
-            lp_reserves >= 0
-        ), f"pricing_models.calc_lp_in_given_tokens_out: ERROR: expected lp_reserves >= 0, not {lp_reserves}!"
+        assert market_state.share_reserves >= 0, (
+            "pricing_models.calc_lp_in_given_tokens_out: ERROR: "
+            f"Expected share_reserves >= 0, not {market_state.share_reserves}!"
+        )
+        assert market_state.bond_reserves >= 0, (
+            "pricing_models.calc_lp_in_given_tokens_out: ERROR: "
+            f"Expected bond_reserves >= 0, not {market_state.bond_reserves}!"
+        )
+        assert market_state.base_buffer >= 0, (
+            "pricing_models.calc_lp_in_given_tokens_out: ERROR: "
+            f"Expected base_buffer >= 0, not {market_state.base_buffer}!"
+        )
+        assert market_state.lp_reserves >= 0, (
+            "pricing_models.calc_lp_in_given_tokens_out: ERROR: "
+            f"Expected lp_reserves >= 0, not {market_state.lp_reserves}!"
+        )
         assert rate >= 0, f"pricing_models.calc_lp_in_given_tokens_out: ERROR: expected rate >= 0, not {rate}!"
-        assert 1 > time_remaining >= 0, (
+        assert 1 > time_remaining.normalized_time >= 0, (
             "pricing_models.calc_lp_in_given_tokens_out: ERROR: "
-            f"expected 1 > time_remaining >= 0, not {time_remaining}!"
+            f"expected 1 > time_remaining >= 0, not {time_remaining.normalized_time}!"
         )
-        assert stretched_time_remaining >= 0, (
+        assert time_remaining.stretched_time >= 0, (
             "pricing_models.calc_lp_in_given_tokens_out: ERROR: "
-            f"expected stretched_time_remaining >= 0, not {stretched_time_remaining}!"
+            f"expected stretched_time_remaining >= 0, not {time_remaining.stretched_time}!"
         )
-        assert share_price >= init_share_price >= 1, (
+        assert market_state.share_price >= market_state.init_share_price >= 1, (
             "pricing_models.calc_lp_in_given_tokens_out: ERROR: "
-            f"expected share_price >= init_share_price >= 1, not, {share_price}!"
+            "expected share_price >= init_share_price >= 1, not "
+            f"share_price={market_state.share_price}, and init_share_price={market_state.init_share_price}"
         )
-        d_shares = d_base / share_price
-        lp_in = (d_shares * lp_reserves) / (share_reserves - base_buffer / share_price)
+        d_shares = d_base / market_state.share_price
+        lp_in = (d_shares * market_state.lp_reserves) / (market_state.share_reserves - market_state.base_buffer)
         # TODO: Move this calculation to a helper function.
-        d_bonds = (share_reserves - d_shares) / 2 * (
-            init_share_price * (1 + rate * time_remaining) ** (1 / stretched_time_remaining) - share_price
-        ) - bond_reserves
+        d_bonds = (market_state.share_reserves - d_shares) / 2 * (
```

Comment:
Case in point of my earlier comment about why this logic should be abstracted. It is tightly coupled to pricing models, so perhaps it does make sense for the two models to live side by side 🤔.